pkg/controller/pxc/upgrade.go

```diff
@@ -185,61 +186,111 @@ func (r *ReconcilePerconaXtraDBCluster) smartUpdate(sfs api.StatefulApp, cr *api
 			primaryPod = pod
 		} else {
 			log.Info(fmt.Sprintf("apply changes to secondary pod %s", pod.Name))
-			if err := r.applyNWait(cr, sfs.StatefulSet().Status.UpdateRevision, &pod, waitLimit); err != nil {
+			if err := r.applyNWait(cr, sfs.StatefulSet(), &pod, waitLimit); err != nil {
 				return fmt.Errorf("failed to apply changes: %v", err)
 			}
 		}
 	}
 
 	log.Info(fmt.Sprintf("apply changes to primary pod %s", primaryPod.Name))
-	if err := r.applyNWait(cr, sfs.StatefulSet().Status.UpdateRevision, &primaryPod, waitLimit); err != nil {
+	if err := r.applyNWait(cr, sfs.StatefulSet(), &primaryPod, waitLimit); err != nil {
 		return fmt.Errorf("failed to apply changes: %v", err)
 	}
 
 	return nil
 }
 
-func (r *ReconcilePerconaXtraDBCluster) applyNWait(cr *api.PerconaXtraDBCluster, updateRevision string, pod *corev1.Pod, waitLimit int) error {
-	if pod.ObjectMeta.Labels["controller-revision-hash"] == updateRevision {
+func (r *ReconcilePerconaXtraDBCluster) applyNWait(cr *api.PerconaXtraDBCluster, sfs *appsv1.StatefulSet, pod *corev1.Pod, waitLimit int) error {
+	if pod.ObjectMeta.Labels["controller-revision-hash"] == sfs.Status.UpdateRevision {
 		log.Info(fmt.Sprintf("pod %s is already updated", pod.Name))
 	} else {
 		if err := r.client.Delete(context.TODO(), pod); err != nil {
 			return fmt.Errorf("failed to delete pod: %v", err)
 		}
 	}
 
-	if err := r.waitPodRestart(updateRevision, pod, waitLimit); err != nil {
+	if err := r.waitPodRestart(sfs.Status.UpdateRevision, pod, waitLimit); err != nil {
 		return fmt.Errorf("failed to wait pod: %v", err)
 	}
 
 	if err := r.waitPXCSynced(cr, pod.Status.PodIP, waitLimit); err != nil {
 		return fmt.Errorf("failed to wait pxc sync: %v", err)
 	}
 
+	if err := r.waitUntilOnline(cr, pod, sfs.Name, waitLimit); err != nil {
+		return fmt.Errorf("failed to wait pxc status: %v", err)
+	}
+
 	return nil
 }
 
-func (r *ReconcilePerconaXtraDBCluster) getPrimaryPod(cr *api.PerconaXtraDBCluster) (string, error) {
+func (r *ReconcilePerconaXtraDBCluster) waitUntilOnline(cr *api.PerconaXtraDBCluster, pod *corev1.Pod, sfsName string, waitLimit int) error {
+	database, err := r.proxyDB(cr)
+	if err != nil {
+		return fmt.Errorf("failed to get proxySQL db: %v", err)
+	}
+
+	defer database.Close()
+
+	fullName := fmt.Sprintf("%s.%s.%s.svc.cluster.local", pod.Name, sfsName, cr.Namespace)
```

Comment:
we cannot guarantee `svc.cluster.local` postfix 😞

Reply:
https://github.com/percona/percona-xtradb-cluster-operator/pull/457/commits/c2dfb30f145c4962a75374c6bdd642960460c54c